### PR TITLE
CI test for dependencies to have the same version

### PR DESCRIPTION
**Pull request by xianny**
## Description

Add failing CI if there are dependencies with multiple versions.
Can specify dependencies or packages to ignore in /package.json

## Testing instructions

Checkout branch, run `yarn remove_node_modules && yarn install`. Then try to build and test everything.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/monorepo-scripts/src/deps_versions.ts [hunk 1]
```diff
@@ -17,11 +17,31 @@ interface VersionsByDependency {
...
+const ignorePackages: string[] = config.ignoreDependencyVersionsForPackage.split(' ');
+
+if (require.main === module) {
+    const dependencies = parseDependencies();
+    const ignoredMultiples = getDependenciesWithMultipleVersions(dependencies.ignored);
+    const multiples = getDependenciesWithMultipleVersions(dependencies.included);
+    printVersionsByDependency(multiples);
+    utils.log(`├── ${chalk.bold('IGNORED PACKAGES')}`);
+    printVersionsByDependency(ignoredMultiples);
+    if (!(Object.keys(multiples).length === 0)) {
```
Comment (fabioberger) [inline packages/monorepo-scripts/src/deps_versions.ts]:
```suggestion
    if (Object.keys(multiples).length !== 0) {
```

File: packages/monorepo-scripts/src/deps_versions.ts [hunk 2]
```diff
@@ -17,11 +17,31 @@ interface VersionsByDependency {
...
 }
 
-const PACKAGE_JSON_GLOB = '../*/package.json';
+interface ParsedDependencies {
+    ignored: VersionsByDependency;
+    included: VersionsByDependency;
+}
+
+const PACKAGE_JSON_GLOB = '../../*/package.json';
+const config = JSON.parse(fs.readFileSync(path.join(__dirname, '../../../package.json')).toString()).config;
```
Comment (fabioberger) [inline packages/monorepo-scripts/src/deps_versions.ts]:
```suggestion

const config = JSON.parse(fs.readFileSync(path.join(__dirname, '../../../package.json')).toString()).config;
```

Let's add a new line between top-of-file constants and the script itself.

File: packages/monorepo-scripts/src/deps_versions.ts [hunk 3]
```diff
@@ -17,11 +17,31 @@ interface VersionsByDependency {
...
+const ignore: string[] = config.ignoreDependencyVersions.split(' ');
+const ignorePackages: string[] = config.ignoreDependencyVersionsForPackage.split(' ');
```
Comment (fabioberger) [inline packages/monorepo-scripts/src/deps_versions.ts]: Can we optimize for readability over variable name shortness? `ignore` -> `dependenciesWithIgnoredVersions` and `ignorePackages` -> `packagesWithIgnoredDependecyVersions`. It's a mouth-full, and I understand why you chose to go with shorter names but I think this does improve the readability. Open to alternative naming proposals.

File: packages/instant/package.json [hunk 4]
```diff
@@ -72,10 +72,10 @@
         "@types/enzyme": "^3.1.14",
         "@types/enzyme-adapter-react-16": "^1.0.3",
         "@types/jest": "^23.3.5",
-        "@types/lodash": "^4.14.116",
+        "@types/lodash": "4.14.104",
         "@types/node": "*",
-        "@types/react": "^16.4.16",
-        "@types/react-dom": "^16.0.8",
+        "@types/react": "*",
+        "@types/react-dom": "*",
```
[the remark this answers is not in the document]
Comment (fragosti) [reply packages/instant/package.json]: I don't think it makes sense for an app to have * as react and react-dom versions. Packages should do that or use `peerDependencies`

File: packages/monorepo-scripts/src/deps_versions.ts [hunk 5]
```diff
@@ -17,11 +17,31 @@ interface VersionsByDependency {
...
+
+if (require.main === module) {
```
[the remark this answers is not in the document]
Comment (xianny) [reply packages/monorepo-scripts/src/deps_versions.ts]: can remove 👍

File: packages/instant/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (xianny) [reply packages/instant/package.json]: should I revert this change?

[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/instant/package.json]: Yes please Xianny. We want `*` for React & React-dom deps within libraries, but not within the app itself. The point of `*` is for the app, and not the dependency to choose the desired version.

**Comment by fabioberger** [comment]
@xianny He want's to point to a specific branch, not to the `master` branch of `@0xproject/react-highlight` (which is our fork of that library.)

**Comment by xianny** [comment]
yep, gotcha! I was wondering why we don't use master - do we keep a different state indefinitely on the branches on that repo? Also, looks like that branch got deleted. I'm pointing to master to fix CI for now. I think this is OK but @fragosti please let me know if you need us to do something different.

**Comment by xianny** [comment]
works locally but fails on CI 🤔

**Comment by fragosti** [comment]
From the build it looks like the types for React might be the wrong version? Not sure.

**Comment by xianny** [comment]
Yeah :/ React versions in instant. I reverted versions to what they were before this PR, and it still fails. The last time it worked was with `@types/react:*` and `@types/react-dom:*`.

**Pull request merged**